### PR TITLE
Return descriptive 503 when listing stake pools with uncomplete sync

### PR DESCRIPTION
**Pull request by hasufell**
Wrt #1971

### CODE REVIEW

File: lib/core/src/Cardano/Wallet.hs [hunk 1]
```diff
@@ -2316,6 +2319,13 @@ data ErrWithdrawalNotWorth
     = ErrWithdrawalNotWorth
     deriving (Generic, Eq, Show)
 
+-- | Errors that can occur when trying to list stake pool.
```
Comment (jonathanknowles) [inline lib/core/src/Cardano/Wallet.hs:2360]: There's more than just one stake pool, I hope! :smile_cat: 

```suggestion
-- | Errors that can occur when trying to list stake pools.
```

File: lib/core/src/Cardano/Wallet.hs [hunk 2]
```diff
@@ -2316,6 +2319,13 @@ data ErrWithdrawalNotWorth
...
+data ErrListPools
+    = ErrListPoolsNetworkError ErrNetworkUnavailable
+    | ErrListPoolsPastHorizonException PastHorizonException
+    deriving (Show)
+
+
```
Comment (jonathanknowles) [inline lib/core/src/Cardano/Wallet.hs]: Some extra space has invaded here:
```suggestion

```
https://github.com/input-output-hk/adrestia/wiki/Coding-Standards#use-only-a-single-blank-line-between-top-level-definitions

File: lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs [hunk 3]
```diff
@@ -184,17 +190,18 @@ newStakePoolLayer nl db@DBLayer {..} =
         :: EpochNo
         -- Exclude all pools that retired in or before this epoch.
         -> Coin
-        -> ExceptT ErrNetworkUnavailable IO [Api.ApiStakePool]
+        -> ExceptT ErrListPools IO [Api.ApiStakePool]
     _listPools currentEpoch userStake = do
         tip <- withExceptT fromErrCurrentNodeTip $ currentNodeTip nl
-        rawLsqData <- stakeDistribution nl tip userStake
+        rawLsqData <- mapExceptT (fmap (first ErrListPoolsNetworkError))
+                        $ stakeDistribution nl tip userStake
```
Comment (jonathanknowles) [inline lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs]: **Request:** use 4-space indents.

```suggestion
        rawLsqData <- mapExceptT (fmap (first ErrListPoolsNetworkError))
            $ stakeDistribution nl tip userStake
```

File: lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs [hunk 4]
```diff
@@ -204,11 +211,16 @@ newStakePoolLayer nl db@DBLayer {..} =
...
+
       where
-        fromErrCurrentNodeTip :: ErrCurrentNodeTip -> ErrNetworkUnavailable
+
+        fromErrCurrentNodeTip :: ErrCurrentNodeTip -> ErrListPools
         fromErrCurrentNodeTip = \case
-            ErrCurrentNodeTipNetworkUnreachable e -> e
-            ErrCurrentNodeTipNotFound -> ErrNetworkUnreachable "tip not found"
+            ErrCurrentNodeTipNetworkUnreachable e -> ErrListPoolsNetworkError e
+            ErrCurrentNodeTipNotFound -> ErrListPoolsNetworkError $ ErrNetworkUnreachable "tip not found"
```
Comment (jonathanknowles) [inline lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs]: **Request:** wrap code to meet the [line length limit](https://github.com/input-output-hk/adrestia/wiki/Coding-Standards#limit-line-length-to-80-characters).

**Justification:** Long lines make it harder to read multiple code sources side-by-side.

```suggestion
            ErrCurrentNodeTipNetworkUnreachable e ->
                ErrListPoolsNetworkError e
            ErrCurrentNodeTipNotFound ->
                ErrListPoolsNetworkError $ ErrNetworkUnreachable "tip not found"
```

File: lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs [hunk 5]
```diff
@@ -204,11 +211,16 @@ newStakePoolLayer nl db@DBLayer {..} =
                 (nOpt rawLsqData)
                 lsqData
                 dbData
+        case r of
+            Left e@(PastHorizon{}) -> throwE (ErrListPoolsPastHorizonException e)
+            Right r' -> pure r'
```
Comment (jonathanknowles) [inline lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs:224]: **Suggestion**: use the `first` combinator to avoid pattern matching on `Right`.

```suggestion
        handlePastHorizon r
      where
        handlePastHorizon = except . first
            (\e@PastHorizon {} -> ErrListPoolsPastHorizonException e)
```

**Review by jonathanknowles — CHANGES_REQUESTED** [review]
Looks good so far. :+1: 

I've added some suggestions in the comments.

There are still some issues with formatting to address. (See comments.)

File: lib/core/src/Cardano/Wallet.hs (hunk 2 above)
Comment (hasufell) [reply lib/core/src/Cardano/Wallet.hs]: As I'm not counting whitespaces, I'm also not counting newlines between code blocks. So we're back to: use a formatter :)

File: lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs (hunk 5 above)
Comment (hasufell) [reply lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs:224]: why?

File: lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs (hunk 3 above)
Comment (hasufell) [reply lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs]: This already is 4-spaced indent.

File: lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs (hunk 5 above)
Comment (jonathanknowles) [reply lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs:224]: > why?

BTW, this was just a suggestion, rather than a request. If you feel the original is clearer, then feel free to leave it that way.

My reasons for suggesting it:

* Using `first` allows you to omit the `Right` pattern match, which means that you can avoid the extra `r'` binding.
* The `first` and `second` functions (from `Data.Bifunctor`) are used fairly commonly in our code base, so it's a familiar pattern to the team.

Comment (hasufell) [reply lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs:224]: In this case it doesn't really matter, but I'd like to point out that these expressions aren't strictly the same:

1. pattern matching forces evaluation immediately, so there are slightly different strictness properties
2. depending on how you define the utility function `handlePastHorizon` it may or may not inline and (at worst) cause an additional closure allocation (in this case I believe it does inline, unless some mtl-ism blocks it from doing so)

This is why I'm usually a bit more conservative with rewriting expressions, because even if they look the same, they aren't always.

**Comment by piotr-iohk** [comment]
@hasufell I have taken your branch (`hasufell:hasufell/1971/fix-err-list-stakepools`) for a little spin and I actually still see the same behavior as -> https://github.com/input-output-hk/cardano-wallet/issues/1971#issuecomment-683685247 :thinking: 

Still see 500 and "Somenthing went wrong" on listing stake pools while wallet is syncing through Byron era. (Checked on [testnet](https://hydra.iohk.io/build/3854628/download/1/index.html))

(btw, is there a reason why you are using your fork of cardano-wallet instead of using main repo?)

**Comment by hasufell** [comment]
I can reproduce. This is the problem as described in https://github.com/input-output-hk/cardano-wallet/issues/1971#issuecomment-683728271

It's hard to say where this exception leaks to.

**Comment by hasufell** [comment]
@piotr-iohk should be fixed, please retry

File: lib/core/src/Cardano/Wallet/Api/Server.hs [hunk 6]
```diff
@@ -1580,16 +1582,18 @@ assignMigrationAddresses addrs selections =
 -------------------------------------------------------------------------------}
 
 data ErrCurrentEpoch = ErrUnableToDetermineCurrentEpoch
+                     | ErrCurrentEpochPastHorizonException PastHorizonException
```
Comment (jonathanknowles) [inline lib/core/src/Cardano/Wallet/Api/Server.hs]: @hasufell This style is different from our team coding standard ([source](https://github.com/input-output-hk/adrestia/wiki/Coding-Standards#avoid-variable-length-indentation)). Please could you fix it?

We use this style in the code base (newlines with 4-space indents):
```suggestion
data ErrCurrentEpoch
    = ErrUnableToDetermineCurrentEpoch
    | ErrCurrentEpochPastHorizonException PastHorizonException
```

Thank-you.

**Review by jonathanknowles — DISMISSED** [review]
Hi @hasufell

There are still some issues with coding style.

In a previous review, I made these requested changes, and they were marked as resolved, but they don't seem to be have been fixed (or if they were, perhaps the fixes were reverted by a subsequent push?):
* https://github.com/input-output-hk/cardano-wallet/pull/2094/files#r480718996
* https://github.com/input-output-hk/cardano-wallet/pull/2094/files#r480723300
* https://github.com/input-output-hk/cardano-wallet/pull/2094/files#r480716728

Please could you address these issues?

As I [mentioned before](https://github.com/input-output-hk/cardano-wallet/pull/2055#discussion_r476343085), I would love to be able to give you a tool that could automatically apply the correct format, but I don't have one. If you'd like to explore automated solutions such as [brittany](https://hackage.haskell.org/package/brittany), I would not be opposed. But that requires a team discussion which is outside the scope of this PR.

Given that our team does not yet use an automated tool, team members generally make a best effort to make their code follow the standard. Could you do that? I don't think it's too hard to follow given that it's rather short, and only has a few rules. We provide an `.editor` file to make this easier, so the burden should not be too arduous.

As a reminder:
* The purpose of having a written standard is to record things that the team has agreed on, so discussions about style can be avoided during PR reviews.
* The benefit is that we can concentrate on more important issues during PR reviews, such as functionality and correctness.
* Rules are added to the standard as the result of team discussions and voting. The standard is a consensus that we've worked out over time.

We allow ourselves the flexibility to occasionally break our own standard if there is good reason. But I don't think that this flexibility should be interpreted as licence to not follow the standard (on the basis that there is no automated tool), especially when the reviewer has gone to the effort of providing suggestions and feedback.

Thank-you.

**Comment by hasufell** [comment]
Hi @jonathanknowles ,

thank you very much for your detailed explanation!

I believe there must be a misunderstanding though. I wasn't arguing against the coding standard in any of my PRs I believe, nor have I indicated that I don't intend to follow it. In fact I've spent a great deal of time configuring my editor, filing bug reports on various vim plugins and reading through stylish-haskell documentation in order to figure out if some of this can be automated (because we already use this tool). I'm using the `.editorconfig` since the first week, but it doesn't do much.

Maybe you're faster at pattern recognition and adapting to different coding styles. I personally am not and I've never worked with this coding style, so I'm afraid it will take more time before my PRs will satisfy your standards, unless this will be automated. Please bear with me.

Thanks for your time reviewing these points!

**Comment by piotr-iohk** [comment]
> @piotr-iohk should be fixed, please retry

LGTM.

```
GET http://localhost:8090/v2/stake-pools?stake=1000000000

The response was:
Code = 503,
Message = {"code":"past_horizon","message":"Tried to convert something that is past the horizon (due to uncertainty about the next hard fork). Wait for the node to finish syncing to the hard fork. Depending on the blockchain, this process can take an unknown amount of time."}
```

**Review by KtorZ — APPROVED** [review]

**Review by piotr-iohk — APPROVED** [review]
